lib/ui/home/home_view_model.dart

```diff
@@ -0,0 +1,81 @@
+import 'package:flutter_riverpod/flutter_riverpod.dart';
+import 'package:flutter_survey/ui/home/survey_ui_model.dart';
+import 'package:rxdart/subjects.dart';
+
+import '../../model/survey.dart';
+import '../../usecases/base/base_use_case.dart';
+import '../../usecases/get_surveys_use_case.dart';
+import 'home_state.dart';
+
+const _defaultPageSize = 5;
+const _loadMoreThreshold = 2;
+
+class HomeViewModel extends StateNotifier<HomeState> {
+  final GetSurveysUseCase _getSurveysUseCase;
+
+  HomeViewModel(
+    this._getSurveysUseCase,
+  ) : super(const HomeState.init()) {
+    loadSurveys();
+  }
+
+  int _page = 1;
+  int _lastSelectedIndex = 0;
+
+  final BehaviorSubject<List<SurveyUiModel>> _surveysSubject =
+      BehaviorSubject();
+
+  Stream<List<SurveyUiModel>> get surveysStream => _surveysSubject.stream;
+
+  final BehaviorSubject<int> _surveyPageIndexSubject = BehaviorSubject();
+
+  Stream<int> get surveyPageIndexStream => _surveyPageIndexSubject.stream;
+
+  void loadSurveys({bool shouldRefresh = false}) async {
+    if (shouldRefresh) {
+      _page = 1;
+      _surveyPageIndexSubject.add(0);
+    }
+    final result = await _getSurveysUseCase.call(GetSurveysInput(
+      pageNumber: _page,
+      pageSize: _defaultPageSize,
+    ));
+    if (result is Success<List<Survey>>) {
+      final uiModels = result.value
+          .map((survey) => SurveyUiModel.fromSurvey(survey))
+          .toList();
+      if (_surveysSubject.hasValue && !shouldRefresh) {
+        _surveysSubject.add(_surveysSubject.value + uiModels);
+      } else {
+        _surveysSubject.add(uiModels);
+      }
+      state = const HomeState.success();
+    } else {
+      _handleError(result as Failed);
+    }
+  }
+
+  void loadMoreSurveys(int selectedIndex) {
```

Comment:
Just curious why do we need to pass a `selectedIndex` here? 🤔

Reply:
@markgravity for two reasons:
1. in case the user swipes back to a previous index beyond the threshold, we don't want to load more. To do that, we keep track of the last selected index.
2. we update the page number before fetching surveys.